### PR TITLE
Introduce `String.join` function

### PR DESCRIPTION
**Pull request by darkdrag00nv2**
Work towards #2752

## Description

Adds `join(_ strs: [String], _ separator: String): String` function to Cadence. The function returns a string value after concatenating elements of the provided array of string with the given separator.
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by SupunS — APPROVED** [review]
Nice!

File: runtime/interpreter/value_string.go [hunk 1]
```diff
@@ -108,6 +108,73 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
...
+	} else {
+		separator = NewStringValue(
+			inter,
+			common.MemoryUsage{
+				Kind:   common.MemoryKindStringValue,
+				Amount: 1,
+			},
+			func() string {
+				return ","
+			})
```
[the remark this answers is not in the document]
Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go]: Moved outside and removed metering.

File: runtime/interpreter/value_string.go [hunk 2]
```diff
@@ -108,6 +108,73 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
...
+	common.UseMemory(inter,
+		common.MemoryUsage{
+			Kind:   common.MemoryKindStringValue,
+			Amount: 1,
+		},
+	)
+	var builder strings.Builder
+	first := true
+
+	argument.Iterate(inter, func(element Value) (resume bool) {
```
[the remark this answers is not in the document]
Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go:147]: Good suggestion, done.

**Review by SupunS — APPROVED** [review]

File: runtime/interpreter/value_string.go [hunk 3]
```diff
@@ -108,6 +108,74 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
...
+		// Add separator
+		if !first {
+			common.UseMemory(inter,
+				common.MemoryUsage{
+					Kind:   common.MemoryKindStringValue,
+					Amount: uint64(len(separator.Str)),
+				},
```
Comment (turbolent) [inline runtime/interpreter/value_string.go:154]: Use `NewStringMemoryUsage` instead. Also please use it in `stringFunctionFromCharacters` and `TypeValue.GetMember` while at it.

File: runtime/interpreter/value_string.go [hunk 4]
```diff
@@ -108,6 +108,74 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
...
+		str, ok := element.(*StringValue)
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+
+		common.UseMemory(inter,
+			common.MemoryUsage{
+				Kind:   common.MemoryKindStringValue,
+				Amount: uint64(len(str.Str)),
+			},
```
Comment (turbolent) [inline runtime/interpreter/value_string.go:169]: Same as above, use `NewStringMemoryUsage`

File: runtime/sema/string_type.go [hunk 5]
```diff
@@ -37,6 +37,11 @@ const StringTypeFromCharactersFunctionDocString = `
 Returns a string from the given array of characters
 `
 
+const StringTypeJoinFunctionName = "join"
+const StringTypeJoinFunctionDocString = `
+Returns a string after joining the array of strings with the given separator.
```
Comment (turbolent) [inline runtime/sema/string_type.go:42]: Also document the default separator if none is given

File: runtime/sema/string_type.go [hunk 6]
```diff
@@ -298,3 +310,23 @@ var StringTypeFromCharactersFunctionType = &FunctionType{
...
+	Parameters: []Parameter{
+		{
+			Label:      ArgumentLabelNotRequired,
+			Identifier: "strs",
+			TypeAnnotation: NewTypeAnnotation(&VariableSizedType{
+				Type: StringType,
+			}),
+		},
+		{
+			Label:          ArgumentLabelNotRequired,
```
Comment (turbolent) [inline runtime/sema/string_type.go:324]: Require an argument label, to make the purpose clear:

```suggestion
```

**Review by turbolent — COMMENTED** [review]
Nice!

**Comment by turbolent** [comment]
@SupunS Contributors can't add labels themselves, please add one for the release notes

File: runtime/interpreter/value_string.go (hunk 3 above)
Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go]: Done, do note that in some places in `stringFunctionJoin` and `stringFunctionFromCharacters`, I had to do a `-1` to account for double counting of empty string.

File: runtime/interpreter/value_string.go [hunk 7]
```diff
@@ -108,6 +100,62 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
...
+	// NewStringMemoryUsage already accounts for empty string.
+	common.UseMemory(inter, common.NewStringMemoryUsage(0))
+	var builder strings.Builder
+	first := true
+
+	argument.Iterate(inter, func(element Value) (resume bool) {
+		// Add separator
+		if !first {
+			// -1 to offset the double counting for empty string.
+			common.UseMemory(inter, common.NewStringMemoryUsage(len(separator.Str)))
```
Comment (SupunS) [inline runtime/interpreter/value_string.go:139]: comment mentions a `-1`, but the implementation doesn't actually decrement.
Same for line#150

File: runtime/interpreter/value_string.go [hunk 8]
```diff
@@ -108,6 +100,62 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
 	return NewUnmeteredStringValue(builder.String())
 }
 
+var StringTypeJoinDefaultSeparator = NewUnmeteredStringValue(",")
+
+func stringFunctionJoin(invocation Invocation) Value {
+	argument, ok := invocation.Arguments[0].(*ArrayValue)
```
Comment (SupunS) [inline runtime/interpreter/value_string.go:106]: maybe rename this variable `argument` to a more intuitive one like `stringArray` or  even just `array`

File: runtime/interpreter/value_string.go [hunk 9]
```diff
@@ -84,22 +84,14 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
...
 	argument.Iterate(inter, func(element Value) (resume bool) {
 		character := element.(CharacterValue)
-		common.UseMemory(inter,
-			common.MemoryUsage{
-				Kind:   common.MemoryKindStringValue,
-				Amount: uint64(len(character)),
-			},
-		)
+		// -1 to offset the double counting for empty string.
+		common.UseMemory(inter, common.NewStringMemoryUsage(len(character)-1))
```
Comment (SupunS) [inline runtime/interpreter/value_string.go:94]: It looks a bit odd to decrement here (though I understand the reason). Maybe that's probably why the original code did it without the constructor. Maybe keep it in the original way, but adding a comment explaining why it is done this way might be better?
cc: @turbolent

Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go]: Personally, I prefer how it was written before. Processing why the extra `-1` is used could take a few seconds and can be avoided in this case. I've updated the code and added the clarification comment.

File: runtime/interpreter/value_string.go (hunk 8 above)
Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go]: Done.

File: runtime/interpreter/value_string.go (hunk 7 above)
Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go]: Reverted back to manual creation and added a comment. The `-1` seems odd.

File: runtime/interpreter/value_string.go (hunk 9 above)
Comment (turbolent) [reply runtime/interpreter/value_string.go]: Yeah, whatever is clearest is best 👍

**Review by SupunS — APPROVED** [review]

File: runtime/interpreter/value_string.go [hunk 10]
```diff
@@ -108,6 +106,74 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
 	return NewUnmeteredStringValue(builder.String())
 }
 
+var StringTypeJoinDefaultSeparator = NewUnmeteredStringValue(",")
```
Comment (turbolent) [inline runtime/interpreter/value_string.go:109]: It's probably better to have the separator default to an empty string
```suggestion
var StringTypeJoinDefaultSeparator = EmptyString
```

File: runtime/sema/string_type.go [hunk 11]
```diff
@@ -37,6 +37,12 @@ const StringTypeFromCharactersFunctionDocString = `
 Returns a string from the given array of characters
 `
 
+const StringTypeJoinFunctionName = "join"
+const StringTypeJoinFunctionDocString = `
+Returns a string after joining the array of strings with the optional separator if provided.
+Uses ',' as the default separator when not provided.
```
Comment (turbolent) [inline runtime/sema/string_type.go:43]:
```suggestion
Uses an empty string as the default separator when not provided.
```

**Review by turbolent — COMMENTED** [review]
Looks good!

Regarding the default: Every language's standard library seems to have a different default, so maybe choosing the empty string as the "lowest common denominator" is best, as it avoids surprises based on wrong expectations.

Maybe we should just require it, i.e. not have any default? 

For example:
- Swift: `""` (empty string)
- JavaScript: `","`
- Kotlin: `", "` (additional space)
- Python, Rust, Go: no default

**Comment by darkdrag00nv2** [comment]
I think it is probably best to just make it required. Made it required in b92d5e1a5.

If we want to have a default, then I would prefer having the same default in `split` and `join`. Empty string won't work in `split`, so I don't prefer it.

File: runtime/interpreter/value_string.go (hunk 10 above)
Comment (darkdrag00nv2) [reply runtime/interpreter/value_string.go]: Not applicable.

File: runtime/sema/string_type.go (hunk 11 above)
Comment (darkdrag00nv2) [reply runtime/sema/string_type.go]: Not applicable.

**Review by dsainati1 — APPROVED** [review]
If I understand correctly, this function doesn't mutate the input array, so making a note to myself to add the appropriate `view` annotation as per https://github.com/onflow/cadence/issues/2466

**Comment by darkdrag00nv2** [comment]
@SupunS @turbolent Should we merge this?

File: runtime/sema/string_type.go [hunk 12]
```diff
@@ -298,3 +310,21 @@ var StringTypeFromCharactersFunctionType = &FunctionType{
...
+				Type: StringType,
+			}),
+		},
+		{
+			Identifier:     "separator",
+			TypeAnnotation: NewTypeAnnotation(StringType),
+		},
+	},
+	ReturnTypeAnnotation: NewTypeAnnotation(StringType),
+	Arity:                &Arity{Min: 2, Max: 2},
```
Comment (turbolent) [inline runtime/sema/string_type.go:329]:
```suggestion
```

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**